src/com/jsyn/io/AudioFifo.java

```diff
@@ -70,62 +75,76 @@ public void close() {
     @Override
     public double read() {
         double value = Double.NaN;
-        if (readWaitEnabled) {
+        if (readWaitEnabled) {         
+            lock.lock();
             try {
-                while (available() < 1) {
-                    synchronized (writeSemaphore) {
-                        writeSemaphore.wait();
-                    }
+              while (available() < 1) {
+                try {
+                    notEmpty.await();
+                } catch (InterruptedException e) {
+                    return Double.NaN;
                 }
-                value = readOneInternal();
-            } catch (InterruptedException e) {
+              }
+              value = readOneInternal();
+            } finally {
+              lock.unlock();
             }
+            
         } else {
             if (readIndex != writeIndex) {
                 value = readOneInternal();
             }
         }
+        
+        if (writeWaitEnabled) {
```

Comment:
Ah good question!
It's mostly a matter of style to separate outer object layers dealing with synchronization primitives from internal layers that don't know anything about locking and are mostly optimized for speed.
It keeps the internal APIs simple, and concentrates the locking/notification logic and mechanism in only one layer of the object, so if you ever need to change the inner layer, you don't have to think about "am I doing the right thing with the locks in here?", and when you do have to deal with the locking logic, you only have to look at one function and never wonder if there is any locking code happening deeper in the stack that you don't "see" directly.